cmd/operator/main.go

```diff
@@ -16,46 +16,123 @@ package main
 
 import (
 	"context"
+	"flag"
 	"os"
+	"strconv"
+	"strings"
+	"time"
 
 	"github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/run"
 	"github.com/GoogleCloudPlatform/cloud-run-release-operator/pkg/config"
 	"github.com/GoogleCloudPlatform/cloud-run-release-operator/pkg/rollout"
 	stackdriver "github.com/TV4/logrus-stackdriver-formatter"
 	isatty "github.com/mattn/go-isatty"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
 	log "github.com/sirupsen/logrus"
 )
 
+type stepFlags []int64
+
+func (steps *stepFlags) Set(step string) error {
+	value, err := strconv.ParseInt(step, 10, 64)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse step")
+	}
+	*steps = append(*steps, value)
+	return nil
+}
+
+func (steps stepFlags) String() string {
+	var value string
+	for _, step := range steps {
+		value += " " + strconv.FormatInt(step, 10)
+	}
+
+	return value
+}
+
+var (
+	flCLI         bool
+	flConfigFile  string
+	flHTTPAddr    string
+	flProject     string
+	flRegion      string
+	flService     string
+	flSteps       stepFlags
+	flStepsString string
+	flInterval    int64
+)
+
+func init() {
+	flag.BoolVar(&flCLI, "cli", false, "run as CLI application to manage rollout in intervals")
+	flag.StringVar(&flHTTPAddr, "http-addr", "", "listen on http portrun on request (e.g. :8080)")
+	flag.StringVar(&flProject, "project", "", "project in which the service is deployed")
+	flag.StringVar(&flRegion, "region", "", "the Cloud Run region where the service is deployed")
+	flag.StringVar(&flService, "service", "", "the service to manage")
+	flag.Var(&flSteps, "step", "a percentage in traffic the candidate should go through")
+	flag.StringVar(&flStepsString, "steps", "", "define steps in one flag separated by commas (e.g. 5,30,60)")
+	flag.Int64Var(&flInterval, "interval", 0, "the time between each rollout step")
+	flag.Parse()
+
+	if !flCLI && flHTTPAddr == "" {
+		log.Fatal("one of -cli or -http-addr must be set")
+	}
```

Comment:
might be wiser to validate these in `main()` I think, `init()`s should be as small as possible.

also in `main()` you'll have a proper logger initialized hopefully.

Reply:
That sounds like good idea. I moved all the validation to `main` and put them all together. I left the piece that obtains the `--steps` in init since that's more about getting a value from the command, which is what the `init` is doing.